spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/KafkaBackoffAwareMessageListenerAdapter.java

```diff
@@ -64,15 +64,32 @@ public KafkaBackoffAwareMessageListenerAdapter(AcknowledgingConsumerAwareMessage
 	}
 
 	public KafkaBackoffAwareMessageListenerAdapter(AcknowledgingConsumerAwareMessageListener<K, V> adapter,
```

Comment:
Ah! Now I see the disconnect about the switch - the other adapters filter, retry, etc can be used with non-POJO listeners (of any type), not just with `@KafkaListener` infrastructure.

In this case we are requiring the delegate to be an ACAML so we don't need the switch.

Do you think it would never be the case that someone would want to wrap their own `MessageListener` in this adapter? I suppose that's reasonable, but we should clearly document it.

Reply:
Hmmm, I actually didn't give much thought to uses besides with the annotation - I myself have never used it in another way. I'll look into it. I think I still don't fully understand the question though, because the MessageListener interface also has all 4 onMessage methods, so I guess the user should be aware of its own MessageListener implementation and handle all 4 methods properly, according to the intended result. I'll give more thought on that later.

But do I think the user should have the flexibility to use this adapter just as any of the others.

Reply:
The hierarchy of interfaces is designed so that they are all `@FunctionalInterface`s, allowing them to be used in lambda expressions:

```java
properties.setMessageListner((MessageListener<String, String) record -> { ... });

...

properties.setMessageListner((ConsumerAwareMessageListener<String, String) (record, consumer) -> { ... });
```

etc.

So, no, users don't have to implement all 4 methods;

Reply:
Oh, I see, everything makes sense now 🙏🏼. Thanks a lot for your patience!